### PR TITLE
codeintel: Rewrite resolvers package

### PR DESCRIPTION
**Pull request by efritz**
Major refactor of the codeintel resolvers package. Fixes both https://github.com/sourcegraph/sourcegraph/issues/15945 and https://github.com/sourcegraph/sourcegraph/issues/15943.

This gets rid of the dependency on the code intel API package from the resolvers package by inlining it. A separate pass will remove now dead code.

This ensures that we are making all calls to gitserver/postgres/codeintel-db on the same (call hierarchy) level so it will be easier to notice where we can omit and batch requests. This should not change any current behavior of the GraphQL API, but may be more efficient due to a more efficient use of database queries.

Major changes:

- *exists.go* was taken almost wholesale from the api package.
- Definitions, Hover, References, Diagnostics, and Ranges resolvers have been removed from *query.go* and put into their own file. All of the new code should be straightforward to follow. It's best to read this new code as new code and not compare it to previous behavior.

### CODE REVIEW

File: enterprise/cmd/frontend/internal/codeintel/resolvers/query_ranges.go [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+		return nil, err
+	}
+
+	for i := range adjustedUploads {
+		ranges, err := r.lsifStore.Ranges(
+			ctx,
+			adjustedUploads[i].Upload.ID,
+			adjustedUploads[i].AdjustedPathInBundle,
+			startLine, // TODO - adjust these as well
+			endLine,   // TODO - adjust these as well
```
Comment (Strum355) [inline enterprise/cmd/frontend/internal/codeintel/resolvers/query_ranges.go:45]: New PR or this PR?

Comment (efritz) [reply enterprise/cmd/frontend/internal/codeintel/resolvers/query_ranges.go:45]: New, this was never handled.

File: enterprise/cmd/frontend/internal/codeintel/resolvers/exists.go [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+// exact document path are returned. Otherwise, dumps containing any document for which the given
+// path is a prefix are returned. These dump IDs should be subsequently passed to invocations of
+// Definitions, References, and Hover.
+func (r *resolver) findClosestDumps(ctx context.Context, cachedCommitChecker *cachedCommitChecker, repositoryID int, commit, path string, exactPath bool, indexer string) (_ []store.Dump, err error) {
+	candidates, err := r.inferClosestUploads(ctx, repositoryID, commit, path, exactPath, indexer)
+	if err != nil {
+		return nil, err
+	}
+
+	filtered := candidates[:0]
```
Comment (Strum355) [inline enterprise/cmd/frontend/internal/codeintel/resolvers/exists.go:32]: image

File: enterprise/cmd/frontend/internal/codeintel/resolvers/query_references.go [hunk 3]
```diff
@@ -0,0 +1,530 @@
...
+	"github.com/sourcegraph/sourcegraph/enterprise/internal/codeintel/stores/dbstore"
+	store "github.com/sourcegraph/sourcegraph/enterprise/internal/codeintel/stores/dbstore"
+	"github.com/sourcegraph/sourcegraph/enterprise/internal/codeintel/stores/lsifstore"
+	"github.com/sourcegraph/sourcegraph/internal/observation"
+)
+
+const slowReferencesRequestThreshold = time.Second
+
+// References returns the list of source locations that reference the symbol at the given position.
+func (r *queryResolver) References(ctx context.Context, line, character, limit int, rawCursor string) (_ []AdjustedLocation, _ string, err error) {
```
Comment (Strum355) [inline enterprise/cmd/frontend/internal/codeintel/resolvers/query_references.go:24]: Just pointing out that I appreciate the comments dotted throughout this method :comfy:

File: enterprise/cmd/frontend/internal/codeintel/resolvers/resolver.go [hunk 4]
```diff
@@ -109,7 +124,7 @@ func (r *resolver) IndexConfiguration(ctx context.Context, repositoryID int) ([]
 	}
 
 	var indented bytes.Buffer
-	json.Indent(&indented, marshaled, "", "\t")
+	_ = json.Indent(&indented, marshaled, "", "\t")
 	return indented.Bytes(), nil
```
Comment (Strum355) [inline enterprise/cmd/frontend/internal/codeintel/resolvers/resolver.go:128]:
```suggestion
	return indented.Bytes(), json.Indent(&indented, marshaled, "", "\t")
```

Add if you want, im impartial :slightly_smiling_face:

**Review by Strum355 — APPROVED** [review]
wide load incoming :shipit:

File: enterprise/cmd/frontend/internal/codeintel/resolvers/query_references.go [hunk 5]
```diff
@@ -0,0 +1,530 @@
...
+}
+
+// ErrConcurrentModification occurs when a page of a references request cannot be resolved as
+// the set of visible uploads have changed since the previous request for the same result set.
+var ErrConcurrentModification = errors.New("result set changed while paginating")
+
+// adjustedUploadsFromCursor adjusts the current target path and the given position for each upload
+// visible from the current target commit. If an upload cannot be adjusted, it will be omitted from
+// the returned slice. The returned slice will be cached on the given cursor. If this data is already
+// stashed o the given cursor, the result is recalculated from the cursor data/resolver context, and
```
Comment (tjdevries) [inline enterprise/cmd/frontend/internal/codeintel/resolvers/query_references.go:119]: `stashed o` <- typo

File: enterprise/cmd/frontend/internal/codeintel/resolvers/resolver.go (hunk 4 above)
Comment (efritz) [reply enterprise/cmd/frontend/internal/codeintel/resolvers/resolver.go:128]: I actually want to do this all the time...

File: enterprise/cmd/frontend/internal/codeintel/resolvers/query_references.go (hunk 5 above)
Comment (efritz) [reply enterprise/cmd/frontend/internal/codeintel/resolvers/query_references.go:119]:
```suggestion
// stashed on the given cursor, the result is recalculated from the cursor data/resolver context, and
```

File: enterprise/cmd/frontend/internal/codeintel/resolvers/resolver.go (hunk 4 above)
Comment (Strum355) [reply enterprise/cmd/frontend/internal/codeintel/resolvers/resolver.go:128]: ikr?

File: enterprise/cmd/frontend/internal/codeintel/resolvers/query_references_cursor.go [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+// calculate the offset into the result set to be returned by the current request.
+type referencesCursor struct {
+	AdjustedUploads           []cursorAdjustedUpload           `json:"adjustedUploads"`
+	DefinitionUploadIDs       []int                            `json:"definitionUploadIDs"`
+	DefinitionUploadIDsCached bool                             `json:"definitionUploadIDsCached"`
+	OrderedMonikers           []lsifstore.QualifiedMonikerData `json:"orderedMonikers"`
+	RemotePhase               bool                             `json:"remotePhase"`
+	LocalOffset               int                              `json:"localOffset"`
+	LocalBatchOffset          int                              `json:"localBatchOffset"`
+	BatchIDs                  []int                            `json:"batchIDs"`
```
Comment (tjdevries) [inline enterprise/cmd/frontend/internal/codeintel/resolvers/query_references_cursor.go:20]: Not sure if it matters, but I always get confused when I just have `IDs` sitting alone. I'm not sure what the batch of IDs are? (maybe I just don't know -- but it's not obvious from looking at the struct for me)

Comment (efritz) [reply enterprise/cmd/frontend/internal/codeintel/resolvers/query_references_cursor.go:20]: I can comment these fields in more detail (or at least point to the function that consumes it).

Comment (tjdevries) [reply enterprise/cmd/frontend/internal/codeintel/resolvers/query_references_cursor.go:20]: Yeah, that'd be helpful. Comment is fine with me.

**Comment by tjdevries** [comment]
In a future PR, do you think it would make sense to create something like a `CursorQueryResolver` ? I just see a ton of `FromCursor` and I could see myself forgetting to use that or something else in the future -- but I'm not sure if that makes sense to think about in Go or if there's a different method I should be thinking.

**Comment by efritz** [comment]
@tjdevries 

The `r.xFromCursor(ctx, cursor, ...)` methods are just constructors that use information in the cursor (if available) and falls back to querying the first-page state of things. I'm not sure this is a generally applicable approach to resolvers (this is really the only resolver that pages with an offset - most of our other endpoints are only controlled by limit).

If you have a sketch of what `CursorQueryResolver` could be like, I'd be curious to see it (but it's not obvious to me what it would look like).

File: enterprise/cmd/frontend/internal/codeintel/resolvers/query_references_cursor.go (hunk 6 above)
Comment (efritz) [reply enterprise/cmd/frontend/internal/codeintel/resolvers/query_references_cursor.go:20]: I'm not ignoring this, but I have stacked work on top of this I want to get merged before I rework this cursor definition a bit.

**Pull request merged**